### ISSUE TITLE
Allow redefinition/shadowing of variable identifier in same scope?

### ISSUE THREAD

**Issue by BryantLam** [chapel-lang/chapel#14106]
Feature Request. Not urgent or anything.

Would it make sense to allow variable identifiers to be reused by new variable declarations in the same scope? Occasionally, I find myself needing to use a new variable identifier despite that the original identifier isn't needed anymore, but cannot be easily moved into a separate scope to allow shadowing.

This comes up more often with nilable types, but is not exclusive to nilable use cases. I append `_` to nilable identifiers when I know I'll borrow them into non-nilables later.

```chapel
class MyClass {
  var x: int;
}

proc main() {
  var arr: [1..8] owned MyClass?;

  var firstElem_ = arr[1];
  if firstElem_ == nil 
    /* handle nil */
  }
  var firstElem = firstElem_!;

  /* Ok to use firstElem from now on. firstElem_ isn't needed anymore. */
}
```

**Comment by BryantLam** [comment]
None of these work today. I'd propose that they could.
```chapel
class MyClass {
  var x: int;
}

proc main() {
  var arr: [1..4] owned MyClass?;
  arr[1] = new owned MyClass(42);

  var firstElem = arr[1];
  if firstElem == nil {
    writeln("handle nil");
  }
  {
    var firstElem = firstElem!; // 1
    writeln(firstElem);
  }

  var firstElem = firstElem!; // 2
  writeln(firstElem);

  for elem in arr {
    if elem == nil { writeln("handle nil"); }
    var elem = elem!; // 3
    writeln(elem);
  }
}
```

**Comment by mppf** [comment]
I would have expected case 1 above to work, actually; but not case 2. I'd expect case 3 to work or not work the same as case 1.

**Comment by lydia-duncan** [comment]
Personally, I would find this style confusing, especially if it would allow you to use a different type than the original.  It's easy miss redefinitions when glancing through larger scopes, making code harder to understand and maintain.

**Comment by bradcray** [comment]
I believe these are all currently working as intended.  Recall that Chapel was designed to resolve symbols in a way that is neutral to declaration order within a scope, so in a case 1-style declaration like `var x = x;` the RHS `x` reference refers to the variable being defined on that same statement.

Case 2 I find most worrisome in terms of confusion / ambiguity.  I don't even think C permits variables to be redefined in this way, does it?

**Comment by BryantLam** [comment]
> Personally, I would find this style confusing, especially if it would allow you to use a different type than the original. It's easy miss redefinitions when glancing through larger scopes, making code harder to understand and maintain.

> Case 2 I find most worrisome in terms of confusion / ambiguity. I don't even think C permits variables to be redefined in this way, does it?

Correct, C does not allow this behavior. In reflecting today, I could live with the current multiple-definition behavior because Chapel has a lot of variable kinds (`var`, `const`, `ref`, ...) that would compound this issue. While a feature like this would also help with code refactoring, I do agree with Lydia that it makes (quickly) scanning the code potentially harder.

**Comment by mppf** [comment]
I'm closing this issue because, in the discussion, it doesn't look like we'd want to do the change proposed here. Additionally, it would be a pretty big change to how the compiler thinks about scoping.
